--- a/include/petscstring.h
+++ b/include/petscstring.h
@@ -286,13 +286,14 @@
 
   PetscFunctionBegin;
   if (!t) PetscFunctionReturn(PETSC_SUCCESS);
-  PetscAssert(n, PETSC_COMM_SELF, PETSC_ERR_ARG_SIZ, "String buffer length must be positive");
+  PetscCheck(n, PETSC_COMM_SELF, PETSC_ERR_ARG_SIZ, "String buffer length must be positive");
   PetscCall(PetscStrlen(s, &len));
+  PetscCheck(len < n, PETSC_COMM_SELF, PETSC_ERR_ARG_SIZ, "Existing string length must be less than buffer capacity");
   PETSC_SILENCE_WSTRINGOP_TRUNCATION_BEGIN;
 #if PetscHasBuiltin(__builtin_strncat)
-  __builtin_strncat(s, t, n - len);
-#else
-  strncat(s, t, n - len);
+  __builtin_strncat(s, t, n - len - 1);
+#else
+  strncat(s, t, n - len - 1);
 #endif
   PETSC_SILENCE_WSTRINGOP_TRUNCATION_END;
   s[n - 1] = '\0';
